### PR TITLE
Kernel driver

### PR DESCRIPTION
**Pull request by diamondman**
First version of the Panda Kernel Driver. 
Works with all 3 CNA interfaces of the Panda. 
Passes workbench test of panda to panda communication.

### CODE REVIEW

File: driver/panda.c [hunk 1]
```diff
@@ -0,0 +1,612 @@
+/**
+ * @file    panda.c
+ * @author  Jeddy Diamond Exum
```
Comment (geohot) [inline driver/panda.c]: lol, typo in your name

File: driver_test/Makefile [hunk 2]
```diff
@@ -0,0 +1,2 @@
+all:
```
Comment (geohot) [inline driver_test/Makefile]: Can we move this to driver/linux/test

File: driver/.gitignore [hunk 3]
```diff
@@ -0,0 +1,6 @@
+.*.cmd
```
Comment (geohot) [inline driver/.gitignore]: Can we move this to driver/linux/

Expecting a windows driver soon

File: driver/panda.c [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+static struct usb_driver panda_usb_driver = {
+  .name = PANDA_MODULE_NAME,
+  .probe = panda_usb_probe,
+  .disconnect = panda_usb_disconnect,
+  .id_table = panda_usb_table,
+};
+
+module_usb_driver(panda_usb_driver);
+
+MODULE_LICENSE("GPL");
```
Comment (geohot) [inline driver/panda.c]: Any issue with "Dual MIT/GPL"?

File: driver/panda.c [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+
+  buf = usb_alloc_coherent(priv_dev->udev, PANDA_USB_RX_BUFF_SIZE,
+			   GFP_KERNEL, &urb->transfer_dma);
+  if (!buf) {
+    dev_err(priv_dev->dev, "No memory left for USB buffer\n");
+    usb_free_urb(urb);
+    return -ENOMEM;
+  }
+
+  usb_fill_int_urb(urb, priv_dev->udev,
```
Comment (geohot) [inline driver/panda.c]: Can we call panda_usb_read_int_callback to trigger this to avoid copied code?

File: driver/panda.c [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+
+    pos += sizeof(struct panda_usb_can_msg);
+  }
+
+ resubmit_urb:
+  usb_fill_int_urb(urb, priv_dev->udev,
+		    usb_rcvintpipe(priv_dev->udev, 1),
+		    urb->transfer_buffer, PANDA_USB_RX_BUFF_SIZE,
+		    panda_usb_read_int_callback, priv_dev, 5);
+
```
Comment (geohot) [inline driver/panda.c]: Does URB_NO_TRANSFER_DMA_MAP have to be set here? If not, why is it set for the first usb_fill_int_urb?

File: driver/panda.c (hunk 4 above)
Comment (diamondman) [inline driver/panda.c]: Code largely influenced by an existing kernel driver. GPL is likely required.

**Pull request merged**